File: backend/infraction-service/app/com/radovan/play/services/impl/EurekaRegistrationServiceImpl.scala
package com.radovan.play.services.impl

import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import com.radovan.play.services.EurekaRegistrationService
import jakarta.inject.{Inject, Singleton}
import org.apache.pekko.actor.ActorSystem
import play.libs.ws.{WSClient, WSResponse}

import java.net.InetAddress
import scala.collection.mutable
import scala.concurrent.ExecutionContext
import scala.concurrent.duration._
import scala.jdk.CollectionConverters._

@Singleton
class EurekaRegistrationServiceImpl @Inject() (
                                                wsClient: WSClient,
                                                objectMapper: ObjectMapper,
                                                actorSystem: ActorSystem,
                                                executionContext: ExecutionContext
                                              ) extends EurekaRegistrationService {

  private val EUREKA_SERVER_URL = "http://localhost:8761/eureka/apps"
  private val appName = "infraction-service"
  private val instanceId = s"$appName-01"
  private val port = 9001

  // 🕒 Periodično zakazivanje registracije
  actorSystem.scheduler.scheduleAtFixedRate(
    initialDelay = 0.seconds,
    interval = 30.seconds
  )(() => registerService())(executionContext)

  override def registerService(): Unit = {
    try {
      val hostname = InetAddress.getLocalHost.getHostName
      val ipAddr = InetAddress.getLocalHost.getHostAddress

      println(s"Hostname: $hostname")
      println(s"IP Address: $ipAddr")

      val instanceData = new mutable.LinkedHashMap[String, Any]()
      instanceData.put("instanceId", instanceId)
      instanceData.put("app", appName)
      instanceData.put("hostName", hostname)
      instanceData.put("ipAddr", ipAddr)
      instanceData.put("statusPageUrl", s"http://$ipAddr:$port/info")
      instanceData.put("healthCheckUrl", s"http://$ipAddr:$port/api/health")
      instanceData.put("homePageUrl", s"http://$ipAddr:$port/")
      instanceData.put("vipAddress", appName)
      instanceData.put("secureVipAddress", appName)
      instanceData.put("leaseRenewalIntervalInSeconds", Int.box(30))
      instanceData.put("leaseExpirationDurationInSeconds", Int.box(90))

      val portMap = new mutable.LinkedHashMap[String, Any]()
      portMap.put("$", Int.box(port))
      instanceData.put("port", portMap.asJava)

      val securePortMap = new mutable.LinkedHashMap[String, Any]()
      securePortMap.put("$", Int.box(0))
      instanceData.put("securePort", securePortMap.asJava)

      val metadata = new mutable.LinkedHashMap[String, Any]()
      metadata.put("management.port", Int.box(port))
      instanceData.put("metadata", metadata.asJava)

      val dataCenterInfo = new mutable.LinkedHashMap[String, Any]()
      dataCenterInfo.put("@class", "com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo")
      dataCenterInfo.put("name", "MyOwn")
      instanceData.put("dataCenterInfo", dataCenterInfo.asJava)

      val registrationData = new mutable.LinkedHashMap[String, Any]()
      registrationData.put("instance", instanceData.asJava)

      val jsonPayload: JsonNode = objectMapper.valueToTree(registrationData.asJava)
      val registrationUrl = s"$EUREKA_SERVER_URL/$appName"

      wsClient.url(registrationUrl)
        .addHeader("Content-Type", "application/json")
        .post(jsonPayload)
        .thenAccept(handleResponse)

    } catch {
      case e: Exception =>
        e.printStackTrace()
        throw new RuntimeException("Failed to register service with Eureka", e)
    }
  }

  private def handleResponse(response: WSResponse): Unit = {
    println(s"Eureka server response status: ${response.getStatus}")
    if (response.getStatus == 204 || response.getStatus == 200) {
      println("Service registered successfully!")
    } else {
      System.err.println(s"Failed to register service with Eureka: ${response.getStatusText}")
      System.err.println(s"Response body: ${response.getBody}")
    }
  }
}

File: backend/registration-service/app/com/radovan/play/services/impl/EurekaRegistrationServiceImpl.java
package com.radovan.play.services.impl;

import com.radovan.play.services.EurekaRegistrationService;
import org.apache.pekko.actor.ActorSystem;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import play.libs.ws.WSClient;
import play.libs.ws.WSResponse;
import scala.concurrent.ExecutionContext;
import scala.concurrent.duration.Duration;

import java.net.InetAddress;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@Singleton
public class EurekaRegistrationServiceImpl implements EurekaRegistrationService {

    private static final String EUREKA_SERVER_URL = "http://localhost:8761/eureka/apps";
    private final WSClient wsClient;
    private final ObjectMapper objectMapper;

    @Inject
    public EurekaRegistrationServiceImpl(WSClient wsClient, ObjectMapper objectMapper, ActorSystem actorSystem, ExecutionContext executionContext) {
        this.wsClient = wsClient;
        this.objectMapper = objectMapper;

        // Zakazivanje periodičnog izvršavanja registerService()
        actorSystem.scheduler().schedule(
                Duration.create(0, TimeUnit.SECONDS),  // Početno kašnjenje
                Duration.create(30, TimeUnit.SECONDS), // Period ponavljanja (30 sekundi)
                this::registerService,                 // Metoda koja se izvršava
                executionContext                       // Kontekst izvršavanja
        );
    }

    @Override
    public void registerService() {
        try {
            System.out.println("Starting service registration...");

            // Dinamičko dohvaćanje IP adrese i hostname-a
            String hostname = InetAddress.getLocalHost().getHostName();
            String ipAddr = InetAddress.getLocalHost().getHostAddress();
            System.out.println("Hostname: " + hostname);
            System.out.println("IP Address: " + ipAddr);

            String appName = "registration-service";
            String instanceId = appName + "-01";
            int port = Integer.valueOf(System.getenv("PLAY_PORT"));

            // Kreiranje podataka za registraciju
            Map<String, Object> instanceData = new HashMap<>();
            instanceData.put("instanceId", instanceId);
            instanceData.put("app", appName);
            instanceData.put("hostName", hostname);
            instanceData.put("ipAddr", ipAddr);
            instanceData.put("statusPageUrl", "http://" + ipAddr + ":" + port + "/info");
            instanceData.put("healthCheckUrl", "http://" + ipAddr + ":" + port + "/api/health");
            instanceData.put("homePageUrl", "http://" + ipAddr + ":" + port + "/");
            instanceData.put("vipAddress", appName);
            instanceData.put("secureVipAddress", appName);
            instanceData.put("leaseRenewalIntervalInSeconds", 30);
            instanceData.put("leaseExpirationDurationInSeconds", 90);

            // Dodavanje porta
            Map<String, Object> portMap = new HashMap<>();
            portMap.put("$", port);
            instanceData.put("port", portMap);

            Map<String, Object> securePortMap = new HashMap<>();
            securePortMap.put("$", 0);
            instanceData.put("securePort", securePortMap);

            // Dodavanje metadata
            Map<String, Object> metadata = new HashMap<>();
            metadata.put("management.port", port);
            instanceData.put("metadata", metadata);

            // Dodavanje dataCenterInfo
            Map<String, Object> dataCenterInfo = new HashMap<>();
            dataCenterInfo.put("@class", "com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo");
            dataCenterInfo.put("name", "MyOwn");
            instanceData.put("dataCenterInfo", dataCenterInfo);

            // Glavni JSON objekat sa "instance" ključem
            Map<String, Object> registrationData = new HashMap<>();
            registrationData.put("instance", instanceData);

            // Konverzija u JSONNode
            JsonNode jsonPayload = objectMapper.valueToTree(registrationData);

            // Slanje POST zahteva ka Eureka serveru
            String registrationUrl = EUREKA_SERVER_URL + "/" + appName;

            wsClient.url(registrationUrl)
                    .addHeader("Content-Type", "application/json")
                    .post(jsonPayload)
                    .thenAccept(this::handleResponse);

        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException("Failed to register service with Eureka", e);
        }
    }

    private void handleResponse(WSResponse response) {
        System.out.println("Eureka server response status: " + response.getStatus());
        if (response.getStatus() == 204 || response.getStatus() == 200) {
            System.out.println("Service registered successfully!");
        } else {
            System.err.println("Failed to register service with Eureka: " + response.getStatusText());
            System.err.println("Response body: " + response.getBody());
        }
    }
}